ConfigSpace/hyperparameters.pyx

```diff
@@ -858,6 +882,303 @@ cdef class NormalFloatHyperparameter(FloatHyperparameter):
+        elif log and self.lower <= 0:
+            raise ValueError("Negative lower bound (%f) for log-scale "
+                            "hyperparameter %s is forbidden." %
+                            (self.lower, name))
+
+        self.default_value = self.check_default(default_value)
+
+        if self.log:
+            if self.q is not None:
+                lower = self.lower - (np.float64(self.q) / 2. - 0.0001)
+                upper = self.upper + (np.float64(self.q) / 2. - 0.0001)
+            else:
+                lower = self.lower
+                upper = self.upper
+            self._lower = np.log(lower)
+            self._upper = np.log(upper)
+        else:
+            if self.q is not None:
+                self._lower = self.lower - (self.q / 2. - 0.0001)
+                self._upper = self.upper + (self.q / 2. - 0.0001)
+            else:
+                self._lower = self.lower
+                self._upper = self.upper
+        if self.q is not None:
+            # There can be weird rounding errors, so we compare the result against self.q, see
+            # In [13]: 2.4 % 0.2
+            # Out[13]: 0.1999999999999998
+            if np.round((self.upper - self.lower) % self.q, 10) not in (0, self.q):
+                raise ValueError(
+                    'Upper bound (%f) - lower bound (%f) must be a multiple of q (%f)'
+                    % (self.upper, self.lower, self.q)
+                )
+
+    def __repr__(self) -> str:
+        repr_str = io.StringIO()
+        repr_str.write("%s, Type: BetaFloat, Alpha: %s Beta: %s, Range: [%s, %s], Default: %s" % (self.name, repr(self.alpha), repr(self.beta), repr(self.lower), repr(self.upper), repr(self.default_value)))
+
+        if self.log:
+            repr_str.write(", on log-scale")
+        if self.q is not None:
+            repr_str.write(", Q: %s" % str(self.q))
+        repr_str.seek(0)
+        return repr_str.getvalue()
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        This method implements a comparison between self and another
+        object.
+
+        Additionally, it defines the __ne__() as stated in the
+        documentation from python:
+            By default, object implements __eq__() by using is, returning NotImplemented
+            in the case of a false comparison: True if x is y else NotImplemented.
+            For __ne__(), by default it delegates to __eq__() and inverts the result
+            unless it is NotImplemented.
+
+        """
+        if not isinstance(other, self.__class__):
+            return False
+
+        return (
+            self.name == other.name and
+            self.default_value == other.default_value and
+            self.alpha == other.alpha and
+            self.beta == other.beta and
+            self.log == other.log and
+            self.q == other.q and
+            self.lower == other.lower and
+            self.upper == other.upper
+        )
+
+    def __copy__(self):
```

Comment:
I would prefer types on magic methods, add a return type

Reply:
Sure!

Reply:
Wait, sorry. None of them had return types as of now, so it's be an inconsistency for me to add them to the Beta parameters. Are you sure I should add a return type?